File: lib/models/book.dart
import 'package:intl/intl.dart';

class Book {
  final List<dynamic> authors;
  final String datetime;
  final String isbn;
  final int price;
  final int quantity;
  final String publisher;
  final String status;
  final String title;
  final String thumbnailUrl;

  Book(
      {this.authors,
      this.datetime,
      this.isbn,
      this.price,
      this.publisher,
      this.status,
      this.title,
      this.thumbnailUrl})
      : quantity = 1;

  factory Book.fromJson(dynamic json) {
    return Book(
      authors: json['authors'] as List<dynamic>,
      datetime: (json['datetime'] as String).substring(0, 10),
      isbn: _parseIsbn(json['isbn'] as String),
      price: json['price'] as int,
      publisher: json['publisher'] as String,
      status: json['status'] as String,
      title: json['title'] as String,
      thumbnailUrl: json['thumbnail'] as String,
    );
  }

  String get formattedPrice {
    final formatter = new NumberFormat('#,###');
    return '${formatter.format(price)}원';
  }

  static String _parseIsbn(String isbnString) {
    final List<String> isbns = isbnString.split(' ');
    if (isbns.length > 1) {
      return isbns[1];
    }
    return isbns[0];
  }

  @override
  String toString() {
    return '${this.isbn}/${this.title}/${this.authors.join(', ')}/${this.status}';
  }
}

File: src/services/BookAPIRepository/Book.ts
export class Book {
  title: string;
  author: string;
  datetime: string;
  isbn: string;
  price: number;
  publisher: string;
  status: string;
  thumbnail: string;

  constructor(
    title: string,
    author: string = "",
    datetime: string = "",
    isbn: string = "",
    price: number = 0,
    publisher: string = "",
    status: string = "",
    thumbnail: string = "no-image"
  ) {
    this.title = title;
    this.author = author;
    this.datetime = datetime;
    this.isbn = isbn;
    this.price = price;
    this.publisher = publisher;
    this.status = status;
    this.thumbnail = thumbnail;
  }

  static fromJson(json): Book {
    const datetime: string = new Date(json.datetime)
      .toISOString()
      .substr(0, 10);
    const isbn: string = json.isbn
      .split(" ")
      .reduce((a: string, b: string) => (a.length > b.length ? a : b));
    const author: string =
      json.authors.length > 1 ? `${json.authors[0]} 외` : json.authors[0];

    return new Book(
      json.title,
      author,
      datetime,
      isbn,
      json.price,
      json.publisher,
      json.status,
      json.thumbnail
    );
  }

  static getHeaders(): string[] {
    return ["ISBN", "도서명", "저자", "출판사", "출판일", "판매상태", "정가"];
  }

  toArray(): any[] {
    return [
      this.isbn,
      this.title,
      this.author,
      this.publisher,
      this.datetime,
      this.status,
      this.price,
    ];
  }
}
